scripts/src/java/org/oppia/android/scripts/ci/RetrieveChangedFiles.kt

```diff
@@ -0,0 +1,127 @@
+ *     by compute_changed_files.
+ * - path_to_bucket_name_output_file: path to the file where the file bucket name corresponding to
+ *     this bucket should be printed.
+ * - path_to_file_list_output_file: path to the file where the list of changed files
+ *     corresponding to this bucket should be printed.
+ * - path_to_test_target_list_output_file: path to the file where the list of changed file's test targets
+ *     corresponding to this bucket should be printed.
+ *
+ * Example:
+ *   bazel run //scripts:retrieve_changed_files -- $(pwd) $CHANGED_FILES_BUCKETS_BASE64_ENCODED_PROTO \\
+ *     $(pwd)/file_bucket_name $(pwd)/changed_files $(pwd)/bazel_test_targets
+ */
+fun main(args: Array<String>) {
+  if (args.size < 5) {
+    println(
+      "Usage: bazel run //scripts:retrieve_changed_files --" +
+        " <encoded_proto_in_base64> <path_to_bucket_name_output_file>" +
+        " <path_to_file_list_output_file> <path_to_test_target_list_output_file>"
+    )
+    exitProcess(1)
+  }
+
+  val repoRoot = args[0]
+  val rootDirectory = File(repoRoot).absoluteFile
+  val protoBase64 = args[1]
+  val bucketNameOutputFile = File(args[2])
+  val fileListOutputFile = File(args[3])
+  val fileTestTargetsListOutputFile = File(args[4])
+
+  val testFileExemptionTextProto = "scripts/assets/test_file_exemptions"
+  val testFileExemptionList by lazy {
+    loadTestFileExemptionsProto(testFileExemptionTextProto)
+      .testFileExemptionList
+      .associateBy { it.exemptedFilePath }
+  }
+
+  ScriptBackgroundCoroutineDispatcher().use { scriptBgDispatcher ->
+    val commandExecutor: CommandExecutor =
+      CommandExecutorImpl(
+        scriptBgDispatcher, processTimeout = 5, processTimeoutUnit = TimeUnit.MINUTES
+      )
+
+    val bazelClient = BazelClient(rootDirectory, commandExecutor)
+
+    val changedFilesBucket =
+      ChangedFilesBucket.getDefaultInstance().mergeFromCompressedBase64(protoBase64)
+
+    val changedFilesTestFiles = changedFilesBucket.changedFilesList.flatMap { changedFile ->
+      val exemption = testFileExemptionList[changedFile]
+      if (exemption != null && exemption.testFileNotRequired) {
+        emptyList()
+      } else {
+        findTestFile(rootDirectory, changedFile)
+      }
+    }
+    val changedFilesTestTargets = bazelClient.retrieveBazelTargets(changedFilesTestFiles)
+    val changedFilesTestTargetWithoutSuffix = changedFilesTestTargets.map { it.removeSuffix(".kt") }
+
+    bucketNameOutputFile.printWriter().use { writer ->
+      writer.println(changedFilesBucket.cacheBucketName)
+    }
+
+    fileListOutputFile.printWriter().use { writer ->
+      writer.println(changedFilesBucket.changedFilesList.joinToString(separator = " "))
+    }
+
+    fileTestTargetsListOutputFile.printWriter().use { writer ->
+      writer.println(changedFilesTestTargetWithoutSuffix.joinToString(separator = " "))
+    }
+  }
+}
+
+private fun findTestFile(rootDirectory: File, filePath: String): List<String> {
```

Comment:
The reason the source files are mapped to find the test files is to get the test targets that can be Built prior to running coverages on the tests. **(for Build Oppia Tests)**

**More context:**

(Correct me if wrong),
Though we provide source files as inputs to our run coverage scripts, the coverages are actually performed running the bazel commands on the test targets. 

In the workflow job order we set up bazel -> build the targets (to help with runs not timing out) -> run the script

So, I thought it might be helpful to have the targets built prior to running, as running coverages by itself would be expensive and having it built and run in one single job might cause timeouts. So, I am mapping these source files to test files, retrieving their test targets and passing them to the BUILD OPPIA TESTS job. 

am I doing it right?

Also, I was additionally thinking of adding one more functionality to 
if we have changed files as `**Test.kt` 
then to acquire its test target too. If the above implementation is right would proceed else will skip it then.

Reply:
So just to check, your plan is to:
- Collect all changed source files (to be passed into run_coverage).
- Collect test targets associated with those source files.
- Use the test targets to build tests ahead of running coverage for improved performance.
- Run coverage for each source file.

Is that correct? If so, I have no concerns with that.

However, it seems to be that the script still needs to map from test file back to source file. If a test is changed in a PR in isolation, we also need to re-run coverage for its corresponding source file. That flow seems missing in the current implementation (unless I'm missing something). This is also a very important case to cover, so we should definitely have tests for it (as we make isolated changes to tests all the time, and that could cause changes in coverage that we want to surface to the PR comment thread).

Reply:
Following up with the review comment - https://github.com/oppia/oppia-android/pull/5465/files/4e4c6877a76af9f1e2fdef8413a92268a9becd90#r1711077577

To specifically answer this ->

The flow is almost right, 
- Collect **all changed source and test files** (to be passed into run_coverage) - all `.kt` files.
- step 2 (as above)
- step 3 (as above)
- Should be able to provide source and test files to Run coverage.
- The Run coverage script maps the test files to source files.
- Run coverage for each source file.

The mapping of test to sources happens in the RunCoverage.kt script (run_coverage) as mentioned in the PR 2.2 description as introduced with enabling support for list of files: #5461  

>[Done] The test files too need to be taken into account while considering coverage analysis (mapping test files to their source files and provide it to the script run, also these files need to be figured out even in the compute changed files utility as that will help with building the targets)

This can be seen in action in the last ci run - running code coverage check (code_coverage.yml) [[stack trace]](https://github.com/oppia/oppia-android/actions/runs/10311895866/job/28546379396?pr=5465#step:17:37)

![image](https://github.com/user-attachments/assets/08d68722-1de7-4447-86c2-7b8802d689d9)

---

### Changes I am planning to make

Reason for having the mapping of test files to source file in the run_coverage script is to have it in a centralized spot as both ci will need to have the mapping and if locally a developer tries to provide a test file (even then) this mapping needs to be done, so had it there.

But now I am planning to also have the mapping in the ci too
1. this could help with also retrieving the source's related test target - to help with build
2. could avoid duplicate coverage run as mentioned in the other comment
3. also would have a cleaner approach for ci

Reply:
Added logic: https://github.com/oppia/oppia-android/pull/5465#discussion_r1711545482

Reply:
Ah I see--thanks for clarifying & making the changes!